### PR TITLE
Implement feature to redirect removed games from the store

### LINKED ISSUE #294: Feature suggestion redirect to community hub if store page removed
IT would be nice to get auto redirected to the comunity hub, if the store page has been removed, for example: Alice: Madness Returns.

### PR DESCRIPTION
**Pull request by rafaelgssa**
Resolves #294 

This uses the `webRequest` API to redirect requests to `store.steampowered.com/app` or `store.steampowered.com/sub` that lead to a 302 status.

`optional_permissions` is used, so that:

- Users can update the extension normally, because adding permissions as optional does not disable the extension after it is updated.
- Users can only grant the permissions if they actually use the option.

Some things I had to change:

- Added a `data-dropdown-trigger` attribute to trigger sub options only when a specific value is chosen in a dropdown. This allows for the "Custom Redirect" sub option to be enabled only when "Custom" is selected in the dropdown.
- Added a  `data-permissions` attribute to make sure that the optional permissions required for the option are requested when the option is saved.

Question: I could not find a guide on how to contribute to the repo, do I have to add all of the localizations?

Quick links for testing:

- https://store.steampowered.com/app/91310
- https://store.steampowered.com/app/19680
- https://store.steampowered.com/sub/98980

### CODE REVIEW

File: js/options/options.js [hunk 1]
```diff
@@ -440,20 +440,33 @@ let Options = (function(){
...
+                }
+            } else if (node.type && node.type === "checkbox") {
                 node.checked = value;
 
                 let parentOption = node.closest(".parent_option");
-                if (parentOption) {
+                if (parentOption && !parentOption.dataset.dropdownTrigger) {
                     if (node.id === "stores_all") value = !value;
-                    for (let nextSibling = parentOption.nextElementSibling; nextSibling.classList.contains("sub_option"); nextSibling = nextSibling.nextElementSibling) {
+                    for (let nextSibling = parentOption.nextElementSibling; nextSibling && nextSibling.classList.contains("sub_option"); nextSibling = nextSibling.nextElementSibling) {
```
Comment (rafaelgssa) [inline js/options/options.js]: Had to add this because it results in an error when the sub option is the last element child.

**Comment by MxtOUT** [comment]
>Question: I could not find a guide on how to contribute to the repo, do I have to add all of the localizations?

No, you don't need to do that ;)

**Comment by rafaelgssa** [comment]
The permissions are already being wrapped into an object. `chrome.permissions.request({ permissions, },` is equivalent to `chrome.permissions.request({ permissions: permissions, },`, and `permissions` is an array. So it's probably something else preventing it from working. I tested only on Firefox, should have tested on Chrome as well. Will do so to try to figure out the problem.

**Comment by MxtOUT** [comment]
My bad, missed those curly brackets

**Comment by rafaelgssa** [comment]
Had to add browser detection to fix it for Chrome. I did so by adding a property to the global `Info` object. If this is not desirable, I can make the detection local in the `addWebRequestListener` function, since it's only used by it at the moment, but I thought having it globally was a better idea in case we need to detect the browser in the future.

I thought any change in permissions disabled the extension, but if you're sure that doesn't happen, should I got ahead and remove the optional permissions?

File: js/background/background.js [hunk 2]
```diff
@@ -736,6 +736,76 @@ class Steam {
...
+                case 'custom':
+                    response.redirectUrl = SyncedStorage.get('redirect_removed_custom')
+                        .replace(/\[TYPE]/, type)
+                        .replace(/\[ID]/, id);
+                    break;
+                default:
+                    break;
+            }
+
+            webRequests.push(details.requestId);
```
Comment (MxtOUT) [inline js/background/background.js]: Isn't that a memory leak here?
If you always push to the array, but never remove anything, it will grow infinitely.

File: js/background/background.js [hunk 3]
```diff
@@ -736,6 +736,76 @@ class Steam {
 Steam._dynamicstore_promise = null;
 Steam._supportedCurrencies = null;
 
+// When redirecting to Steam Community, if the game does not actually exist, onHeadersReceived would normally trigger an infinite redirection loop. To avoid that, we store the IDs of the requests that have already been redirected here.
+let webRequests = [];
```
Comment (MxtOUT) [inline js/background/background.js]: A `Set` would suit better here

File: js/background/background.js [hunk 4]
```diff
@@ -736,6 +736,76 @@ class Steam {
...
+    if (details.statusCode === 302 && webRequests.indexOf(details.requestId) < 0) {
+        let matches = details.url.match(/(app|sub)\/(\d+)/);
+
+        if (matches) {
+            let type = matches[1];
+            let id = matches[2];
+
+            switch (SyncedStorage.get('redirect_removed_to')) {
+                case 'community':
+                    response.redirectUrl = `https://steamcommunity.com/${type}/${id}`;
```
Comment (MxtOUT) [inline js/background/background.js]: afaik subs don't have a community page.
Maybe redirect to SteamDB then?
What do you think?

File: js/background/background.js [hunk 5]
```diff
@@ -736,6 +736,76 @@ class Steam {
...
+                case 'steamdb':
+                    response.redirectUrl = `https://steamdb.info/${type}/${id}`;
+                    break;
+                case 'custom':
+                    response.redirectUrl = SyncedStorage.get('redirect_removed_custom')
+                        .replace(/\[TYPE]/, type)
+                        .replace(/\[ID]/, id);
+                    break;
+                default:
+                    break;
```
Comment (MxtOUT) [inline js/background/background.js]: Why is that here?

File: js/background/background.js [hunk 6]
```diff
@@ -736,6 +736,76 @@ class Steam {
...
+            }
+
+            webRequests.push(details.requestId);
+        }
+    }
```
Comment (MxtOUT) [inline js/background/background.js]: You can add an `else` here and delete the entries

File: js/background/background.js [hunk 7]
```diff
@@ -736,6 +736,76 @@ class Steam {
...
+
+function addWebRequestListener() {
+    let filter = {
+        types: ['main_frame'], // Only requests triggered by the user through regular navigation.
+        urls: ['*://*.store.steampowered.com/app/*', '*://*.store.steampowered.com/sub/*'],
+    };
+    let extraInfoSpec = ['blocking'];
+
+    if (Info.browser === 'chrome') {
+        extraInfoSpec.push('extraHeaders'); // Required since Chrome v72.
```
Comment (MxtOUT) [inline js/background/background.js]: Where did you find this?

File: js/background/background.js (hunk 2 above)
Comment (rafaelgssa) [reply js/background/background.js]: I had tried to fix that with `onCompleted`, but it doesn't work well, because we would need permissions for the redirected hosts in order to listen to the completed event for them. So we would need `<all_urls>` permission, and that's a bad idea.

I guess another way to fix it would be to empty the array every once in a while?

File: js/background/background.js (hunk 5 above)
Comment (rafaelgssa) [reply js/background/background.js]: I follow the convention of always adding the default clause in my code, didn't realize you don't follow it. Will remove it.

File: js/background/background.js (hunk 7 above)
Comment (rafaelgssa) [reply js/background/background.js]: https://developer.chrome.com/extensions/webRequest There are a ton of warnings about things that changed on Chrome v72 and Chrome v79 below the event definitions.

File: js/background/background.js (hunk 4 above)
Comment (rafaelgssa) [reply js/background/background.js]: Ok, should I add an option to allow choosing the redirect for apps / subs separately?

File: js/background/background.js (hunk 2 above)
Comment (MxtOUT) [reply js/background/background.js]: We only need this for steamcommunity.com as you already said and we have the permission for this.
In order to work properly you'd have to only add to the `Set` if the user chose to redirect to steamcommunity.com

File: js/background/background.js [hunk 8]
```diff
@@ -736,6 +736,76 @@ class Steam {
 Steam._dynamicstore_promise = null;
 Steam._supportedCurrencies = null;
 
+// When redirecting to Steam Community, if the game does not actually exist, onHeadersReceived would normally trigger an infinite redirection loop. To avoid that, we store the IDs of the requests that have already been redirected here.
+let webRequests = [];
+
+function onHeadersReceived(details) {
+    let response = {};
+
+    if (details.statusCode === 302 && webRequests.indexOf(details.requestId) < 0) {
```
Comment (Revadike) [inline js/background/background.js]: I think using `includes` would be more preferable.

File: js/background/background.js (hunk 7 above)
Comment (MxtOUT) [reply js/background/background.js]: I've seen this, but we're not using any of the headers that this option provides, if I read it correctly

File: js/core.js [hunk 9]
```diff
@@ -869,3 +886,24 @@ function sleep(duration) {
         setTimeout(function() { resolve(); }, duration);
     });
 }
+
+
+class OptionalPermissions {
```
Comment (Revadike) [inline js/core.js]: This class would be useful for other cases too, like context menu. Maybe add a TODO comment to implement it to other areas later.

File: js/background/background.js (hunk 8 above)
Comment (rafaelgssa) [reply js/background/background.js]: I'm switching to a Set.

File: js/background/background.js (hunk 7 above)
Comment (rafaelgssa) [reply js/background/background.js]: I think this is what applies:

```
Starting from Chrome 72, if you need to modify responses before Cross Origin Read Blocking (CORB) can block the response, you need to specify 'extraHeaders' in opt_extraInfpSpec. 
```

Not sure. What I know for sure is that as soon as I added `extraHeaders`, it started redirecting normally on Chrome.

File: js/core.js (hunk 9 above)
Comment (rafaelgssa) [reply js/core.js]: I was actually going to remove this class, as I'll be changing the permissions from optional to required.

**Comment by rafaelgssa** [comment]
Applied the changes.

I made the permissions required, but left the optional permissions logic there, because it could be useful for other features, like @Revadike brought up above. Is that ok?

File: js/background/background.js [hunk 10]
```diff
@@ -736,6 +736,80 @@ class Steam {
...
+            switch (SyncedStorage.get('redirect_removed_to')) {
+                case 'community':
+                    // Subs do not have a community page. In that case, we redirect to SteamDB instead.
+                    if (type === 'app') {
+                        webRequests.add(details.requestId);
+                        response.redirectUrl = `https://steamcommunity.com/${type}/${id}`;
+                        break;
+                    }
+                case 'steamdb':
+                    response.redirectUrl = `https://steamdb.info/${type}/${id}`;
```
Comment (xPaw) [inline js/background/background.js]:
```suggestion
                    response.redirectUrl = `https://steamdb.info/${type}/${id}/`;
```

**Comment by MxtOUT** [comment]
It seems like event pages and the `webRequest` permission don't go well together, I can't load the extension with this.

I'm sorry, but you have to revert this.

Additionally, I'm not super convinced about `OptionalPermissions.remove`:
Why should you remove a permission once you acquired it?
If the user later decides to use that option again, he is greeted with another prompt.
I would just remove this function

**Comment by rafaelgssa** [comment]
Hmm, I didn't notice that Augmented Steam uses event pages instead of background ones, bummer. So do you want me to revert everything except for the optional permissions logic for this PR?

Well, Chrome says this:

> You should remove permissions when you no longer need them. After a permission has been removed, calling permissions.request() usually adds the permission back without prompting the user. 

And Firefox:

> Any permissions granted are retained by the extension, even over upgrade and disable/enable cycling.

So it's best practice to remove them when not needed and the user shouldn't be prompted again after re-enabling the option.

**Comment by MxtOUT** [comment]
I see, but then `remove` is still not perfectly right:
What would happen if multiple options require the same permission?
It would just remove it even though another option may depend on it

**Comment by MxtOUT** [comment]
Maybe we could create a heavily stripped down persistent background page just for the `webRequest` stuff.

I don't know about the performance and whether that's possible at all though

**Comment by rafaelgssa** [comment]
> What would happen if multiple options require the same permission?

Yeah, that would be a problem, I'll see if I can implement a way to only remove the permission if it's not used by any other option.

> Maybe we could create a heavily stripped down persistent background page just for the webRequest stuff.

I don't think that's possible.

**Comment by rafaelgssa** [comment]
The optional permissions logic and the browser detection have been moved to separate PRs (#659, #661).

I'll implement a solution for this using the regular way (https://github.com/tfedor/AugmentedSteam/issues/294#issuecomment-520268922).

Should I just do that in this PR or close this and create a new one, as I'll have to revert everything anyway?

**Comment by MxtOUT** [comment]
Would be better to open a new one

**Comment by Revadike** [comment]
So, what happened @rafaelgomesxyz? As far as I can see, this PR has been split, but its main feature (the redirect for removed games) never made it into a new PR?

**Comment by rafaelgomesxyz** [comment]
Yeah, I ended up forgetting about it. I might pick it up again if I have the time.

**Comment by Revadike** [comment]
This may be interesting to know: @MxtOUT said
> With https://github.com/tfedor/AugmentedSteam/commit/1c8564d270c3d377dd8b15d1318edb9eff49b470 we changed to a permanent background page, so we can properly use the `webRequest` API now

**Comment by alpe12** [comment]
I would love this option. :/
